### PR TITLE
(More) correctly type AirplaneFunc

### PR DESCRIPTION
**Pull request by leeweisberger**
The AirplaneFunc should be parametrized by the parameters that the function uses, not the task config.

This also supports calling the Func with no params if the config has no params or params is an empty object.

I added unit tests that test that the types work correctly and the function is passing params correctly to the execute function.

There is one (unfortunate) hack due to what I think is a TS bug - but I could not get everything working correctly without it!

### CODE REVIEW

**Comment by colinking** [comment]
@leeweisberger I took a look here. +1 on the changes to the function type. +1 also to allowing folks to call tasks without a parameter list if the task **has no required parameters**.

AD: let's move `executeOptions` onto the task function.

If a task has no required parameters, I'd want its signature to be:

```ts
(params?: TParamValues | undefined | null, executeOptions?: ExecuteOptions) => Promise<TOutput>
```

This would let you call this task as `task()` if you don't need to pass parameters or execute options. If you want to pass just execute options, you could do any of `task({}, {host: "..."})` or `task(undefined, {host: "..."})` or `task(null, {host: "..."})`.

If a task has no parameters at all, which signature should it have?

```ts
(params?: {} | undefined | null, executeOptions?: ExecuteOptions) => Promise<TOutput>
// or...
(executeOptions?: ExecuteOptions) => Promise<TOutput>
```

I think the latter makes more sense? LMK what you think.

I also pushed up a small commit that does some renaming. I made those changes while I was reading the code, but lmk if you dislike any of the changes.

**Review by colinking — CHANGES_REQUESTED** [review]
Requesting changes for:
1. Handling the case where a task has only optional parameters.
2. Moving `executeOptions`

If you just want to test types, I'd recommend using https://effectivetypescript.com/2022/05/28/eslint-plugin-expect-type/ instead of Jest.

**Comment by leeweisberger** [comment]
> @leeweisberger I took a look here. +1 on the changes to the function type. +1 also to allowing folks to call tasks without a parameter list if the task **has no required parameters**.
> 
> AD: let's move `executeOptions` onto the task function.
> 
> If a task has parameters, but no required parameters, I'd want its signature to be:
> 
> ```ts
> (params?: TParamValues | undefined | null, executeOptions?: ExecuteOptions) => Promise<TOutput>
> ```
> 
> This would let you call this task as `task()` if you don't need to pass parameters or execute options. If you want to pass just execute options, you could do any of `task({}, {host: "..."})` or `task(undefined, {host: "..."})` or `task(null, {host: "..."})`.
> 
> If a task has no parameters at all, which signature should it have?
> 
> ```ts
> (params?: {} | undefined | null, executeOptions?: ExecuteOptions) => Promise<TOutput>
> // or...
> (executeOptions?: ExecuteOptions) => Promise<TOutput>
> ```
> 
> I think the latter makes more sense? LMK what you think.
> 
> I also pushed up a small commit that does some renaming. I made those changes while I was reading the code, but lmk if you dislike any of the changes.

I got this working for required fields - but you need to call `task({})`. I'm sure it's possible to support `task()`, but I sunk enough time into this already :(

**Review by colinking — APPROVED** [review]

File: packages/airplane/internal/config.test.ts [hunk 1]
```diff
@@ -126,14 +121,47 @@ describe("config", () => {
...
     );
 
-    // TODO this should work
-    // await myTask();
+    mockResponse({ first: "Colin", last: "King" });
+    await myTask({ first: "Colin", last: "King" }, { host });
 
-    myTask({ name: "Colin" });
+    mockResponse({ first: "Colin", last: "King", middle: "Dunno" });
+    await myTask({ first: "Colin", last: "King", middle: "Dunno" }, { host });
```
[the remark this answers is not in the document]
Comment (leeweisberger) [reply packages/airplane/internal/config.test.ts:164]: What is it?

[the remark this answers is not in the document]
Comment (colinking) [reply packages/airplane/internal/config.test.ts:164]: > middle: "Dunno"

**Pull request merged**